From 1924e25730e7d551bdfc5ec373100770b509efe3 Mon Sep 17 00:00:00 2001
From: Mamoru TASAKA <mtasaka@fedoraproject.org>
Date: Fri, 20 Dec 2024 17:03:04 +0900
Subject: [PATCH] Support ruby 3.4 backtrace formatting change

ruby 3.4 changes backtrace formatting, including
opening quote using single quote instead of backtick.

Make DBus::ProxyObject support both backtrace format.

Fixes #144 .
---
 lib/dbus/proxy_object.rb | 12 +++++++-----
 spec/signal_spec.rb      |  3 ++-
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/lib/dbus/proxy_object.rb b/lib/dbus/proxy_object.rb
index f889fe2..e52f575 100644
--- a/lib/dbus/proxy_object.rb
+++ b/lib/dbus/proxy_object.rb
@@ -33,6 +33,8 @@ class ProxyObject
     # @return [ApiOptions]
     attr_reader :api
 
+    OPEN_QUOTE = RUBY_VERSION >= "3.4" ? "'" : "`"
+
     # Creates a new proxy object living on the given _bus_ at destination _dest_
     # on the given _path_.
     def initialize(bus, dest, path, api: ApiOptions::CURRENT)
@@ -58,7 +60,7 @@ def interfaces
     def [](intfname)
       introspect unless introspected
       ifc = @interfaces[intfname]
-      raise DBus::Error, "no such interface `#{intfname}' on object `#{@path}'" unless ifc
+      raise DBus::Error, "no such interface #{OPEN_QUOTE}#{intfname}' on object #{OPEN_QUOTE}#{@path}'" unless ifc
 
       ifc
     end
@@ -127,7 +129,7 @@ def has_iface?(name)
     # @return [void]
     def on_signal(name, &block)
       unless @default_iface && has_iface?(@default_iface)
-        raise NoMethodError, "undefined signal `#{name}' for DBus interface `#{@default_iface}' on object `#{@path}'"
+        raise NoMethodError, "undefined signal #{OPEN_QUOTE}#{name}' for DBus interface #{OPEN_QUOTE}#{@default_iface}' on object #{OPEN_QUOTE}#{@path}'"
       end
 
       @interfaces[@default_iface].on_signal(name, &block)
@@ -151,7 +153,7 @@ def method_missing(name, *args, &reply_handler)
         # - di not specified
         # TODO
         # - di is specified but not found in introspection data
-        raise NoMethodError, "undefined method `#{name}' for DBus interface `#{@default_iface}' on object `#{@path}'"
+        raise NoMethodError, "undefined method #{OPEN_QUOTE}#{name}' for DBus interface #{OPEN_QUOTE}#{@default_iface}' on object #{OPEN_QUOTE}#{@path}'"
       end
 
       begin
@@ -159,10 +161,10 @@ def method_missing(name, *args, &reply_handler)
       rescue NameError => e
         # interesting, foo.method("unknown")
         # raises NameError, not NoMethodError
-        raise unless e.to_s =~ /undefined method `#{name}'/
+        raise unless e.to_s =~ /undefined method #{OPEN_QUOTE}#{name}'/
 
         # BTW e.exception("...") would preserve the class.
-        raise NoMethodError, "undefined method `#{name}' for DBus interface `#{@default_iface}' on object `#{@path}'"
+        raise NoMethodError, "undefined method #{OPEN_QUOTE}#{name}' for DBus interface #{OPEN_QUOTE}#{@default_iface}' on object #{OPEN_QUOTE}#{@path}'"
       end
     end
     # rubocop:enable Lint/MissingSuper
diff --git a/spec/signal_spec.rb b/spec/signal_spec.rb
index 7e449d5..d9b83f6 100755
--- a/spec/signal_spec.rb
+++ b/spec/signal_spec.rb
@@ -107,9 +107,10 @@ def new_quitter(main_loop)
   describe DBus::ProxyObject do
     describe "#on_signal" do
       it "raises a descriptive error when the default_iface is wrong" do
+        OPEN_QUOTE = RUBY_VERSION >= "3.4" ? "'" : "`"
         @obj.default_iface = "org.ruby.NoSuchInterface"
         expect { @obj.on_signal("Foo") {} }
-          .to raise_error(NoMethodError, /undefined signal.*interface `org.ruby.NoSuchInterface'/)
+          .to raise_error(NoMethodError, /undefined signal.*interface #{OPEN_QUOTE}org.ruby.NoSuchInterface'/)
       end
     end
   end
